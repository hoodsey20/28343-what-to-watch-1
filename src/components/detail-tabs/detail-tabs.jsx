import React from 'react';
import PropTypes from 'prop-types';

import withActiveTab from '../../hocs/with-active-tab/with-active-tab';
import withDetailData from '../../hocs/with-detail-data/with-detail-data';

function getRatingName(rating) {
  if (rating < 3) {
    return `Bad`;
  } else if (rating < 5) {
    return `Normal`;
  } else if (rating < 8) {
    return `Good`;
  } else if (rating < 10) {
    return `Very good`;
  }
  return `Awesome`;
}

function getFormattedTime(time) {
  return `${Math.floor(time / 60)}h ${time % 60}m`;
}

function getFormattedDate(dateString) {
  const options = {year: `numeric`, month: `long`, day: `numeric`};
  return new Date(dateString).toLocaleDateString(`en-US`, options);
}

function renderReviews(reviews) {
  return reviews.map((it) => (
    <div key={it.id} className="review">
      <blockquote className="review__quote">
        <p className="review__text">{it.comment}</p>
        <footer className="review__details">
          <cite className="review__author">{it.user.name}</cite>
          <time className="review__date">{getFormattedDate(it.date)}</time>
        </footer>
      </blockquote>

      <div className="review__rating">{it.rating}</div>
    </div>
  ));
}

export const DetailTabs = ({
  movie,
  changeTabHandler,
  activeTab,
  tabs,
  reviews,
}) => {
  const {
    released,
    genre,
    director,
    runTime,
    starring,
    rating,
    scoresCount,
    description,
  } = movie;
  return (
    <div className="movie-card__desc">
      <nav className="movie-nav movie-card__nav">
        <ul className="movie-nav__list">
          {Object.keys(tabs).map((it) => (
            <li
              key={it}
              className={`movie-nav__item ${Number(it) === activeTab ? `movie-nav__item--active` : ``}`}
            >
              <a
                href="#"
                className="movie-nav__link"
                onClick={(evt) => changeTabHandler(evt, Number(it))}
              >
                {tabs[it]}
              </a>
            </li>
          ))}
        </ul>
      </nav>

      {tabs[activeTab] === `Overview` &&
        <React.Fragment>
          <div className="movie-rating">
            <div className="movie-rating__score">{rating}</div>
            <p className="movie-rating__meta">
              <span className="movie-rating__level">{getRatingName(rating)}</span>
              <span className="movie-rating__count">{scoresCount} ratings</span>
            </p>
          </div>

          <div className="movie-card__text">
            <p>{description}</p>
            <p className="movie-card__director"><strong>Director: {director}</strong></p>
            <p className="movie-card__starring"><strong>{starring.join(`, `)} and other</strong></p>
          </div>
        </React.Fragment>
      }

      {tabs[activeTab] === `Details` &&
        <div className="movie-card__text movie-card__row">
          <div className="movie-card__text-col">
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Director</strong>
              <span className="movie-card__details-value">{director}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Starring</strong>
              <span className="movie-card__details-value">
                {starring.map((it, i) => (
                  <span key={i}>
                    {it}
                    <br />
                  </span>
                ))}
              </span>
            </p>
          </div>

          <div className="movie-card__text-col">
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Run Time</strong>
              <span className="movie-card__details-value">{getFormattedTime(runTime)}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Genre</strong>
              <span className="movie-card__details-value">{genre}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Released</strong>
              <span className="movie-card__details-value">{released}</span>
            </p>
          </div>
        </div>
      }

      {(tabs[activeTab] === `Reviews` && !!reviews) &&
        <div className="movie-card__reviews movie-card__row">
          <div className="movie-card__reviews-col">
            {renderReviews(reviews.slice(reviews.length / 2))}
          </div>
          <div className="movie-card__reviews-col">
            {renderReviews(reviews.slice(0, reviews.length / 2))}
          </div>
        </div>
      }
    </div>
  );
};

export default withActiveTab(
    withDetailData(DetailTabs)
);

DetailTabs.propTypes = {
  activeTab: PropTypes.number.isRequired,
  changeTabHandler: PropTypes.func.isRequired,
  tabs: PropTypes.objectOf(PropTypes.string),

  movie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    released: PropTypes.number,
    runTime: PropTypes.number,
    genre: PropTypes.string,
    director: PropTypes.string,
    description: PropTypes.string,
    rating: PropTypes.number,
    scoresCount: PropTypes.number,
    starring: PropTypes.arrayOf(PropTypes.string),
  }),

  reviews: PropTypes.arrayOf(
      PropTypes.shape({
        comment: PropTypes.string,
        date: PropTypes.string,
        id: PropTypes.number,
        rating: PropTypes.number,
        user: PropTypes.shape({
          id: PropTypes.number,
          name: PropTypes.string,
        })
      })
  ),
};
