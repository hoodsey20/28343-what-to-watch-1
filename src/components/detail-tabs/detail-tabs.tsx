import * as React from "react";
import { Movie, Review, TabData } from "../../types";

import withActiveTab from "../../hocs/with-active-tab/with-active-tab";

function getRatingName(rating: number) {
  if (rating < 3) {
    return `Bad`;
  } else if (rating < 5) {
    return `Normal`;
  } else if (rating < 8) {
    return `Good`;
  } else if (rating < 10) {
    return `Very good`;
  }
  return `Awesome`;
}

function getFormattedTime(time) {
  return `${Math.floor(time / 60)}h ${time % 60}m`;
}

function getFormattedDate(dateString) {
  const options = { year: `numeric`, month: `long`, day: `numeric` };
  return new Date(dateString).toLocaleDateString(`en-US`, options);
}

function renderReviews(reviews) {
  return reviews.reverse().map(it => (
    <div key={it.id} className="review">
      <blockquote className="review__quote">
        <p className="review__text">{it.comment}</p>
        <footer className="review__details">
          <cite className="review__author">{it.user.name}</cite>
          <time className="review__date">{getFormattedDate(it.date)}</time>
        </footer>
      </blockquote>

      <div className="review__rating">{it.rating}</div>
    </div>
  ));
}

interface Props {
  movie: Movie;
  reviews: Review[];
  changeTabHandler: (evt: React.SyntheticEvent, id: number) => void;
  activeTab: number;
  tabs: TabData;
}

export const DetailTabs: React.SFC<Props> = ({
  movie,
  changeTabHandler,
  activeTab,
  tabs,
  reviews
}) => {
  const {
    released,
    genre,
    director,
    runTime,
    starring,
    rating,
    scoresCount,
    description
  } = movie;
  return (
    <div className="movie-card__desc">
      <nav className="movie-nav movie-card__nav">
        <ul className="movie-nav__list">
          {Object.keys(tabs).map(it => (
            <li
              key={it}
              className={`movie-nav__item ${
                Number(it) === activeTab ? `movie-nav__item--active` : ``
              }`}
            >
              <a
                href="#"
                className="movie-nav__link"
                onClick={evt => changeTabHandler(evt, Number(it))}
              >
                {tabs[it]}
              </a>
            </li>
          ))}
        </ul>
      </nav>

      {tabs[activeTab] === `Overview` && (
        <React.Fragment>
          <div className="movie-rating">
            <div className="movie-rating__score">{rating}</div>
            <p className="movie-rating__meta">
              <span className="movie-rating__level">
                {getRatingName(rating)}
              </span>
              <span className="movie-rating__count">{scoresCount} ratings</span>
            </p>
          </div>

          <div className="movie-card__text">
            <p>{description}</p>
            <p className="movie-card__director">
              <strong>Director: {director}</strong>
            </p>
            <p className="movie-card__starring">
              <strong>{starring.join(`, `)} and other</strong>
            </p>
          </div>
        </React.Fragment>
      )}

      {tabs[activeTab] === `Details` && (
        <div className="movie-card__text movie-card__row">
          <div className="movie-card__text-col">
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Director</strong>
              <span className="movie-card__details-value">{director}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Starring</strong>
              <span className="movie-card__details-value">
                {starring.map((it, i) => (
                  <span key={i}>
                    {it}
                    <br />
                  </span>
                ))}
              </span>
            </p>
          </div>

          <div className="movie-card__text-col">
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Run Time</strong>
              <span className="movie-card__details-value">
                {getFormattedTime(runTime)}
              </span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Genre</strong>
              <span className="movie-card__details-value">{genre}</span>
            </p>
            <p className="movie-card__details-item">
              <strong className="movie-card__details-name">Released</strong>
              <span className="movie-card__details-value">{released}</span>
            </p>
          </div>
        </div>
      )}

      {tabs[activeTab] === `Reviews` && !!reviews && (
        <div className="movie-card__reviews movie-card__row">
          <div className="movie-card__reviews-col">
            {renderReviews(reviews.slice(reviews.length / 2))}
          </div>
          <div className="movie-card__reviews-col">
            {renderReviews(reviews.slice(0, reviews.length / 2))}
          </div>
        </div>
      )}
    </div>
  );
};

export default withActiveTab(DetailTabs);
