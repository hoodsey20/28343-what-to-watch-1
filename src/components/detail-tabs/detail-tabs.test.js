import React from 'react';
import renderer from 'react-test-renderer';

import {DetailTabs} from './detail-tabs.jsx';

const movieMock = {"name": `Aviator`, "posterImage": `https://es31-server.appspot.com/wtw/static/film/poster/Aviator.jpg`, "previewImage": `https://es31-server.appspot.com/wtw/static/film/preview/aviator.jpg`, "backgroundImage": `https://es31-server.appspot.com/wtw/static/film/background/Aviator.jpg`, "backgroundColor": `#D6CDAF`, "description": `A biopic depicting the early years of legendary Director and aviator Howard Hughes' career from the late 1920s to the mid 1940s.`, "rating": 3.8, "scoresCount": 307174, "director": `Martin Scorsese`, "starring": [`leonardoDiCaprio`, `cateBlanchett`, `kateBeckinsale`], "runTime": 170, "genre": `Drama`, "released": 2014, "id": 2, "isFavorite": false, "videoLink": `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`, "previewVideoLink": `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`};

const reviewsMock = [{"id": 1, "user": {"id": 13, "name": `Zak`}, "rating": 4, "comment": `I really find it difficult to believe this movie is rated highly by people. It's hands down the worst movie I've seen in my life`, "date": `2019-05-27T15:15:34.699Z`}, {"id": 2, "user": {"id": 12, "name": `Isaac`}, "rating": 4, "comment": `The editing is a mess, and the transitions of the plot or characters are rather strange. There is no narrative focus and the storytelling is unbalanced. I cannot really understand why such a bad movie received an overwhelming approval from the critics. `, "date": `2019-06-03T15:15:34.699Z`}];

const mockTabs = {
  0: `Overview`,
  1: `Details`,
  2: `Reviews`,
};

it(`Movie Detail Page is rendered correctly`, () => {
  const tree = renderer.create(<DetailTabs
    movie={movieMock}
    reviews={reviewsMock}
    tabs={mockTabs}
    changeTabHandler={jest.fn()}
    activeTab={0}
  />).toJSON();

  expect(tree).toMatchSnapshot();
});
