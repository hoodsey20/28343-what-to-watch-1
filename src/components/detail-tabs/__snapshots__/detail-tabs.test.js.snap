// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie Detail Page is rendered correctly 1`] = `
<div
  className="movie-card__desc"
>
  <nav
    className="movie-nav movie-card__nav"
  >
    <ul
      className="movie-nav__list"
    >
      <li
        className="movie-nav__item movie-nav__item--active"
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Overview
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Details
        </a>
      </li>
      <li
        className="movie-nav__item "
      >
        <a
          className="movie-nav__link"
          href="#"
          onClick={[Function]}
        >
          Reviews
        </a>
      </li>
    </ul>
  </nav>
  <div
    className="movie-rating"
  >
    <div
      className="movie-rating__score"
    >
      3.8
    </div>
    <p
      className="movie-rating__meta"
    >
      <span
        className="movie-rating__level"
      >
        Normal
      </span>
      <span
        className="movie-rating__count"
      >
        307174
         ratings
      </span>
    </p>
  </div>
  <div
    className="movie-card__text"
  >
    <p>
      A biopic depicting the early years of legendary Director and aviator Howard Hughes' career from the late 1920s to the mid 1940s.
    </p>
    <p
      className="movie-card__director"
    >
      <strong>
        Director: 
        Martin Scorsese
      </strong>
    </p>
    <p
      className="movie-card__starring"
    >
      <strong>
        leonardoDiCaprio, cateBlanchett, kateBeckinsale
         and other
      </strong>
    </p>
  </div>
</div>
`;
